Step3 - 로또(자동)

lewisseo91 opened the pull request:
안녕하세요! 
3단계 로또(자동) 완료 후에 PR 보냅니다.
리뷰 해주셔서 정말 감사합니다!
잘 부탁드립니다~

lewisseo91 committed:
refactor(test): parameterizedTest 적용

합계 계산, 예외 테스트로 나누어 parameterizedTest 시도
- null & 빈문자 입력 테스트
- 커스텀 구분자 없는 합계 테스트
- 커스텀 구분자 있는 합계 테스트
- 여러개 숫자, 여러 구분자 합계 테스트(커스텀 구분자)
- 여러개 숫자, 여러 구분자 합계 테스트
- 예외 음수, 숫자 아닌 입력 예외 테스트

lewisseo91 committed:
refactor(test): DelimitersTest, NumbersTest parameterizedTest 적용

DelimitersTest, NumbersTest parameterizedTest
- 커스텀 구분자 있는 입력, 커스텀 구분자 없이 입력
- 숫자 하나 입력, 숫자 여러개 입력

lewisseo91 committed:
fix(domain): final 추가

number, numbers 불변 객체로 변경

lewisseo91 committed:
refactor(domain): 기본 구분자 입력 방식 변경, 빈문자 or null 확인 상위 이동

기본 구분자 입력 방식 변경, 빈문자 or null 확인 상위 이동
- 기본 구분자 입력 상위에서 입력 받고 Delimiters 는 구분자 처리만 담당
- isNullOrEmpty 하위 처리로 null 값 처리가 너무 많아 상위에서 처리 후 0값 반환

lewisseo91 committed:
refactor(domain): Numbers null 확인 제거

Numbers null 확인 제거
- null 확인이 상위 이동되었기 때문에 정리

lewisseo91 committed:
refactor(domain): delimiters 불변 객체로 설정

String -> StringBuilder 로 변경 후 delimiters final 정의

lewisseo91 committed:
refactor(domain): CalculatorManager 삭제 후 StringAddCalculator 로 이동

StringAddCalculator 역할이 전달 역할밖에 없어 CalculatorManager 삭제 후 상위 이동

lewisseo91 committed:
feat(docs): 로또 기능 목록, 테스트 케이스 작성, 주의 사항 작성

로또 기능 목록, 테스트 케이스 작성, 주의 사항 작성

lewisseo91 committed:
feat(LottoNumberTest): 로또 숫자 테스트 케이스 추가

정답 숫자 1개 내 숫자 1개 테스트 케이스 추가
- 1 - 1 -> true
- 1 - 4 -> false
정답 숫자 3개 내 숫자 1개 테스트 케이스 추가
- 1,2,3 - 1 -> true
- 1,2,3 - 4 -> false

lewisseo91 committed:
feat(LottoNumber): 로또 숫자 추가

로또 숫자 추가
- 1 - 1 -> true
- 1 - 4 -> false
로또 숫자 유효성 확인
- 숫자인지 확인
- 로또 숫자 범위 안인지 확인

lewisseo91 committed:
chore(docs): LottoNumber 숫자 구현 기능 목록 체크

내 숫자가 정답 숫자를 포함하는 지 확인 체크
내 숫자가 정답 숫자를 포함하지 않는 지 확인 체크

lewisseo91 committed:
feat(LottoNumbersTest): 숫자들 테스트 케이스 추가

내 숫자가 정답 숫자에 포함된 경우 테스트
내 숫자들이 6개 다 맞는 경우 테스트
내 숫자들이 0개 맞는 경우 테스트
내 숫자들이 3개 맞는 경우 테스트

lewisseo91 committed:
feat(LottoNumbers): 숫자들 추가

LottoNumbers 추가
- lottoNumbers Set 으로 받는 생성자 추가
- lottoNumbers String 으로 받는 생성자 추가 (수동 입력 있어서)
  - generateLottoNumbers 메서드 추가 (Set 으로 만드는 역할)
- containsNumber 메서드 추가 (contains 체크)
- getMatchCount 메서드 추가 (내 숫자들이 몇 개 맞는 지 확인)

lewisseo91 committed:
chore(docs): 숫자들 기능 목록 체크

- 내 숫자들 묶음이 정답 숫자를 포함하는 지 확인
- 내 숫자들 묶음이 정답 숫자를 몇 개 포함하는 지 확인

lewisseo91 committed:
feat(BuyAmountTest): 구매 개수 테스트 추가

- 가격을 토대로 구매 개수 출력 테스트

lewisseo91 committed:
feat(BuyAmount): 구매 개수 추가

BuyAmount 생성
 - 구매 개수 구하는 메서드(getAmount) 추가
 - 구매 개수를 토대로 가격 구하는 메서드(getPrice) 추가

lewisseo91 committed:
chore(docs): 구매 개수 기능 목록 체크

구매 개수 기능 목록 체크

lewisseo91 committed:
feat(LottoRankingStatusTest): 로또 순위 상태 테스트 추가

현재 match 3개 ~ 6개 까지 있으므로 범위 확인 테스트

lewisseo91 committed:
feat(LottoRankingStatus): 로또 순위 상태 추가 (enum)

- match 3개 ~ 6개 추가
- none 추가
- matchAmount 확인하는 getMatchAmount 메서드 추가

lewisseo91 committed:
chore(docs): 등수 생성 체크

- 등수 생성 체크

lewisseo91 committed:
feat(LottoResultTest): 로또 결과 테스트 추가

- 6개 다 맞는 경우 MATCH6 상태를 가지는 지 확인
- 0개 맞는 경우 NONE 상태를 가지는 지 확인
- 3개 맞는 경우 MATCH3 상태를 가지는 지 확인

lewisseo91 committed:
feat(LottoResult): 로또 결과 추가

getResultRanking 메서드로 matchAmount 개수를 기준으로 로또 상태를 가져오는 메서드 추가

lewisseo91 committed:
chore(docs): 기능 목록 로또 몇등인지 결과 확인 체크

로또 몇등인지 결과 확인 체크

lewisseo91 committed:
feat(LottoResultsTest): 로또 결과들 추가

로또 결과들 중 6개 다 맞춘 순위의 개수 테스트 추가

lewisseo91 committed:
feat(LottoResults): 로또 결과들 추가

- generateLottoRankingMap 메서드로 LottoRankingStatus, 개수를 가지는 Map 생성
- getMatchAmount 메서드로 원하는 LottoRankingStatus 의 개수를 가져옴
- getTotalReward 메서드로 현재 map 의 총 합계를 가져옴
- getProfitRate 메서드로 총 합계를 기준으로 수익률을 구함

lewisseo91 committed:
chore(docs): 등수의 개수를 증가 기능 목록 체크

등수의 개수를 증가 기능 목록 체크

lewisseo91 committed:
feat(LottoResultsViewTest): 로또 결과 뷰(UI) 테스트 추가

현재는 System 테스트를 아직 추가하지 않아 log 만 확인 중
- 로또 결과 확인 showLottoResults 추가
- 로또 총 수익 확인 getLottoResultsRewardTotal 추가
- 로또 수익률 확인 getLottoResultsProfitRate 추가

lewisseo91 committed:
feat(LottoResultsView): 로또 결과 뷰(UI) 추가

LottoResultsView.java 로또 결과 확인, 로또 수익률 확인 추가
- 로또 결과 확인 showResults 추가
- 로또 총 수익 확인 showTotalReward 추가
- 로또 수익률 확인 showProfitRate 추가
LottoResults 로또 결과 확인 시 필요한 toString 추가

lewisseo91 committed:
style(LottoResultsViewTest): 로또 결과 뷰(UI) 이름 추가

로또 결과 뷰 displayName 추가
- 로또 총 수익 확인 getLottoResultsRewardTotal displayName 추가
- 로또 수익률 확인 getLottoResultsProfitRate displayName 추가

lewisseo91 committed:
chore(docs): 로또 결과 뷰(UI) 기능 목록 체크

- 모든 묶음을 확인하면 각 등수 마다 몇 개 있는지 확인
- 등수 개수의 합에 따라 수익률이 얼마인지 확인

lewisseo91 committed:
style(LottoResultsView): showResults 정리

showResults toString 없어도 override 했기 때문에 결과 값이 같으므로 제거

lewisseo91 committed:
feat(LottoNumbersGroupTest): 로또 숫자 묶음 그룹 테스트 추가

- 랜덤 로또 숫자 그룹 생성 테스트 추가
- 로또 숫자 수동 그룹 생성 테스트 추가
  (자동 제어를 아직 못해서 수동으로 추가 후 수정 예정)
- 로또 그룹 결과 테스트 추가

lewisseo91 committed:
feat(LottoNumbersGroup): 로또 숫자 묶음 그룹 추가

LottoNumbersGroup 추가 내용
- 랜덤 로또 숫자 그룹 생성자 추가 (buyAmount, prizeLottoNumbers)
  - generateRandomLottoNumbers 메서드로 생성
- 수동 로또 숫자 그룹 생성자 추가 (String[] myLottoNumbersList, prizeLottoNumbers)
  - generateManualLottoNumbers 메서드로 생성
- 기본 로또 숫자 그룹 생성자 추가 (List<LottoNumbers> myLottoNumbersList, prizeLottoNumbers)
- 로또 그룹 결과 추가 (getLottoResults)
LottoNumbers 추가 내용
- 생성자에 아무것도 통과하지 않을 경우 randomNumbers 생성 하도록 변경
- getLottoResult 메서드 추가 (로또 결과 반환)
- toString 추가
LottoNumber 추가 내용
- toString 추가
- getNumber 추가 (Numbers toString 정렬 용)

lewisseo91 committed:
chore(docs): lottoNumbersGroup 구현 기능 체크

내 숫자를 생성 (숫자들 묶음의 그룹) 체크 완료

lewisseo91 committed:
fix(LottoResultsViewTest): UI 테스트 제거

UI 테스트 제거

lewisseo91 committed:
refactor(LottoNumbersGroupTest): LottoNumbersGroupTest prizeLottoNumbers 제거

LottoNumbersGroupTest prizeLottoNumbers 제거
 기존 constructor(..., LottoNumbers prizeLottoNumbers)
  ->
  constructor(...)
 로 변경
 (꼭 당첨 LottoNumbers 없이도 LottoNumbers는 생성될 수 있기 때문에 변경)

lewisseo91 committed:
refactor(LottoNumbersGroup): LottoNumbersGroup prizeLottoNumbers 제거

LottoNumbersGroup prizeLottoNumbers 제거
 기존 constructor(..., LottoNumbers prizeLottoNumbers)
  ->
  constructor(...)
 로 변경
 (꼭 당첨 LottoNumbers 없이도 LottoNumbers는 생성될 수 있기 때문에 변경)

lewisseo91 committed:
feat(LottoNumbersTest): LottoNumbersTest 당첨 번호 입력 관련 추가

LottoNumbersTest 당첨 번호 입력 관련 추가
- 당첨 로또 입력 테스트 커스텀 구분자 경우 추가
- 당첨 로또 입력 정렬 테스트 (정렬 되지 않은 순서의 입력)

lewisseo91 committed:
fix(LottoResults): LottoResultsView에서 UI 외 로직을 처리하던 부분 이동

LottoResultsView 수익률 구할 때 소수점 처리를 ui 쪽에서 하고 있었는데 LottoResults 에서 버림 후 표시로 변경

lewisseo91 committed:
feat(LottoResultsTest): 로또 결과 수익률, 총 수익 테스트 추가

로또 결과 수익률, 총 수익 테스트 추가

lewisseo91 committed:
style(ExceedNumberBoundException): 예외 directory 생성 후 이동

예외 directory 생성 후 이동

lewisseo91 committed:
feat(LottoResultsView): 결과 UI 추가

당첨 결과 UI 추가
- showResult 에서 resultsTitle, resultsStats, ProfitRate를 보여주도록 추가

lewisseo91 committed:
feat(LottoInputView): 입력 UI 추가

showAskBuyPriceMessage : 구입 금액 요구 메시지
showBuyStats : 구매 후 표시되는 구매 개수와 자신의 로또 번호들
showAskPrizeLottoNumbersMessage : 당첨 번호 입력 메시지

lewisseo91 committed:
feat(LottoMessage): 메시지 필요한 부분들 LottoMessage로 모두 이동

LottoResultsView -> showResultsTitle, showResultsStats, showProfitRate 이동
LottoInputView -> showAskBuyPriceMessage, showBuyAmount, showMyLottoNumbers, showAskPrizeLottoNumbersMessage 이동

lewisseo91 committed:
feat(LottoApplication): 로또 메인 함수 추가

로또 메인 함수 추가

lewisseo91 committed:
feat(LottoController): 로또 컨트롤러 추가

로또 컨트롤러 추가
- startGame 게임 시작
- inputBuyPrice 구입금액 요구 메서드
- showBuyStats 구입 후 구매 개수, 로또 번호들 표시 메서드
- inputPrizeLottoNumbers 정답 로또 입력 요구 메서드
- showResults 결과 통계 표시 메서드
- getScanner 스캐너 새로 받는 메서드

lewisseo91 committed:
feat(LottoNumbers): 정답 로또 입력 시 스페이스 관계없이 받아오게 수정

정답 로또 입력 시 스페이스 관계없이 받아오게 수정
 - replaceAll 사용

lewisseo91 committed:
feat(BuyAmountTest): 구매 오류 테스트 추가

최소 티켓 금액 이하 구매 시도 시 오류 발생하도록 추가
 - 요구사항에는 없었고 자의적으로 필요하다고 생각해서 추가했기에 오류 발생 동작은 변경 가능성 있음

lewisseo91 committed:
feat(MinimumTicketPriceException): 최소 티켓 금액 이하 구매 시도 오류 추가

MinimumTicketPriceException 추가
- 최소 티켓 금액 이하 구매 시도 오류 추가 : runtimeException 상속
BuyAmount validateBuyPrice 추가
- 최소 티켓 금액 이하 구매 시 오류 발생하도록 메서드 추가

lewisseo91 committed:
feat(docs): 중간에 기록못한 테스트 케이스 추가

추가된 항목
- 구매 개수 테스트
- 로또 등수 상태 테스트
- 로또 결과 테스트
- 로또 결과들 테스트
- 로또 숫자들 그룹 테스트

lewisseo91 committed:
chore(): 테스트들 displayName 추가

class 들 위에 대표이름들 추가

lewisseo91 committed:
chore(LottoNumber): 줄 바꿈 정리

LottoNumber
줄 바꿈 정리
LottoResultTest
쓰지 않는 패키지 제거

lewisseo91 committed:
refactor(LottoNumbers): random 번호 생성할 때 static으로 변경

random 번호 생성할 때 static 으로 변경 (한번만 생성하고 셔플을 계속하면 되기 때문)

lewisseo91 committed:
feat(LottoNumbersGroupTest): random 로또 번호 override로 테스트 추가

override 후 객체 주소가 달라져서 assertThat 에 toString 으로 하게 되었는데 mock 으로 하는 방법 필요
하지만 지금은 가능한 방법으로 commit

lewisseo91 committed:
chore(): 줄 바꿈 및 간격 정리

줄 바꿈 및 간격 정리

lewisseo91 committed:
feat(WrongLottoNumberSizeException): 잘못된 정답 로또 개수 입력 시 예외 추가

LottoNumbers
- 잘못된 정답 로또 개수 입력 시 예외 추가
WrongLottoNumberSizeException
- 예외 클래스 생성

lewisseo91 committed:
feat(BuyAmountTest): 구매 개수 예외 테스트 메시지 확인 추가

구매 개수 예외 테스트 메시지 확인 추가

nokchax commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,64 @@
+        this.number = Integer.parseInt(lottoNumber);
+    }
+
+    public int getNumber() {
+        return number;
+    }
```
sorting 때문에 넣으신것 같은데
``` java
public class LottoNumber implements Comparable<LottoNumber>  {
...
```
Comparable 을 구현하시면 불필요한 getter를 제거 할 수 있겠네요

lewisseo91 replied:
> sorting 때문에 넣으신것 같은데
> 
> ```java
> public class LottoNumber implements Comparable<LottoNumber>  {
> ...
> ```
> 
> Comparable 을 구현하시면 불필요한 getter를 제거 할 수 있겠네요

맞습니다 감사합니다~! Comparable 구현하겠습니다~!

nokchax commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,133 @@
+public class LottoNumbers {
+    public static final String FIND_ALL_SPACES = "\\s+";
+    public static final String REMOVE_SPACES = "";
+    public static final String LOTTO_NUMBERS_BASE_SEPARATOR = ",";
+    public static final String LOTTO_NUMBERS_DESCRIPTION_OPEN_BRACKET = "[";
+    public static final String LOTTO_NUMBERS_DESCRIPTION_CLOSE_BRACKET = "]";
+    public static final String LOTTO_NUMBERS_DESCRIPTION_SPACE = " ";
```
view와 관련된 변수들이네요
LottoNumbers 라는 모델 내부에 view와 관련된 것들이 있으면 MVC 패턴에 위배된다고 할 수 있어요
toString(), addElementForMakingString() 등과 같은 함수를 없애고
DTO 를 통해서 출력에 필요한 데이터를 반환하고 ResultView에서 이 dto를 활용해서 출력하도록 변경해보세요

lewisseo91 replied:
> view와 관련된 변수들이네요 LottoNumbers 라는 모델 내부에 view와 관련된 것들이 있으면 MVC 패턴에 위배된다고 할 수 있어요 toString(), addElementForMakingString() 등과 같은 함수를 없애고 DTO 를 통해서 출력에 필요한 데이터를 반환하고 ResultView에서 이 dto를 활용해서 출력하도록 변경해보세요

감사합니다! 참고해서 수정하겠습니다!

nokchax commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,133 @@
+        return resultPocket;
+    }
+
+    private void generateRandomLottoNumbers() {
+        Collections.shuffle(randomLottoNumberPocket);
+        for (int index = 0; index < LOTTO_SIZE; index++) {
+            LottoNumber randomNumber = randomLottoNumberPocket.get(index);
+            lottoNumbers.add(randomNumber);
+        }
+    }
```
```suggestion
    private Set<LottoNumber> generateRandomLottoNumbers() {
        Collections.shuffle(randomLottoNumberPocket);
        
        return new HashSet<>(randomLottoNumberPocket.subList(0, LOTTO_SIZE);
    }
```
subList를 통해 간결하게 표현할 수 있을거 같아요

lewisseo91 replied:
> subList를 통해 간결하게 표현할 수 있을거 같아요

와우! 감사합니다!

nokchax commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,133 @@
+        int lastIndex = sortedLottoNumbers.size() - 1;
+        for (LottoNumber lottoNumber : sortedLottoNumbers) {
+            result.append(lottoNumber.toString());
+            if (index == lastIndex) {
+                break;
+            }
```
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

lewisseo91 replied:
> 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

아이고 감사합니다!

nokchax commented on src/main/java/lotto/LottoNumbers.java:
```diff
@@ -0,0 +1,133 @@
+
+    private List<LottoNumber> convertSetToSortedLottoNumbersList(Set<LottoNumber> lottoNumbers) {
+        return lottoNumbers.stream()
+                .sorted(Comparator.comparingInt(LottoNumber::getNumber))
```
```suggestion
                .sorted()
```
LottoNumber 클래스에 Comparable 인터페이스를 구현하면 간결하게 표현이 가능해요

nokchax commented on src/main/java/lotto/LottoRankingStatus.java:
```diff
@@ -0,0 +1,44 @@
+    MATCH6(6, 2000000000),
+    NONE(0, 0);
+
+    public static final String MATCH_DESCRIPTION = "%s개 일치 (%s원)- %s개";
```
이것 또한 view 와 관련된 코드네요

nokchax commented on src/main/java/lotto/LottoRankingStatus.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Arrays;
+
+public enum LottoRankingStatus {
+    MATCH3(3, 5000),
+    MATCH4(4, 50000),
+    MATCH5(5, 1500000),
+    MATCH6(6, 2000000000),
```
```suggestion
    MATCH3(3, 5_000),
    MATCH4(4, 50_000),
    MATCH5(5, 1_500_000),
    MATCH6(6, 2_000_000_000),
```
이러면 가독성을 높일 수 있습니다.

nokchax commented on src/main/java/lotto/LottoResults.java:
```diff
@@ -0,0 +1,62 @@
+    public int getTotalReward() {
+        return lottoRankingAmountMap.keySet()
+                .stream()
+                .mapToInt(lrs -> lrs.getPrizeReward(lottoRankingAmountMap.get(lrs)))
```
lrs 이러한 약어는 지양해 주세요

nokchax commented on src/main/java/lotto/LottoResults.java:
```diff
@@ -0,0 +1,62 @@
+    public String toString() {
+        StringBuilder result = new StringBuilder();
+        for (LottoRankingStatus lottoRankingStatus : lottoRankingAmountMap.keySet()) {
+            if (lottoRankingStatus == LottoRankingStatus.NONE) {
+                continue;
+            }
```
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

nokchax commented on src/main/java/lotto/ui/LottoController.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private static Scanner getScanner() {
+        return new Scanner(System.in);
```
매번 객체를 생성해 줄 필요가 있을까요?

lewisseo91 replied:
> 매번 객체를 생성해 줄 필요가 있을까요?

제가 scanner input 쪽이 익숙치 않아 계속 생성하는 식으로 했는데
하지 않고 하는 방법 알아보겠습니다 감사합니다!

nokchax commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -1,120 +1,45 @@
 
 public class StringAddCalculatorTest {
+    @ParameterizedTest(name = "{index}. {0} 테스트")
+    @MethodSource("provideStringsForUnitTest")
```
`@MethodSource` 활용 👍

nokchax commented on src/test/java/lotto/LottoNumbersGroupTest.java:
```diff
@@ -0,0 +1,86 @@
+
+    private List<LottoNumbers> getExampleLottoNumbersList() {
+        return Arrays.asList(
+                new LottoNumbers(
+                        Stream.of(new LottoNumber(1), new LottoNumber(2), new LottoNumber(3),
+                                new LottoNumber(4), new LottoNumber(5), new LottoNumber(6))
+                                .collect(Collectors.toSet())),
+                new LottoNumbers(
+                        Stream.of(new LottoNumber(4), new LottoNumber(2), new LottoNumber(3),
+                                new LottoNumber(13), new LottoNumber(12), new LottoNumber(14))
+                                .collect(Collectors.toSet())),
+                new LottoNumbers(
+                        Stream.of(new LottoNumber(42), new LottoNumber(12), new LottoNumber(24),
+                                new LottoNumber(32), new LottoNumber(11), new LottoNumber(15))
+                                .collect(Collectors.toSet()))
```
테스트를 위해 반복되는 코드들은 Fixture를 만들어서 사용하는 것도 좋을거 같아요

nokchax commented on src/main/java/lotto/LottoResults.java:
```diff
@@ -0,0 +1,62 @@
+public class LottoResults {
+    public static final double PROFIT_RATE_DECIMAL_POINT = 100.0;
+    public static final int PROFIT_BASE = 1;
+    private final Map<LottoRankingStatus, Integer> lottoRankingAmountMap;
```
변수명에 map, list 같은 자료구조 형 네이밍은 지양해주세요
코드 전반적으로 보이는데 수정이 필요해 보입니다 😢 

lottoRankingAmounts 정도의 복수형이면 충분할 것 같습니다.

lewisseo91 replied:
> 변수명에 map, list 같은 자료구조 형 네이밍은 지양해주세요 코드 전반적으로 보이는데 수정이 필요해 보입니다 😢
> 
> lottoRankingAmounts 정도의 복수형이면 충분할 것 같습니다.

네 감사합니다! 네이밍 수정하겠습니다~

nokchax reviewed (changes requested):
안녕하세요, 동현님!
세번째 미션도 잘 진행해 주셨습니다 👍
클래스와 메서드를 잘 나누어 설계 하셨어요 :)
몇가지 코멘트 남겼습니다. 확인후 재요청 부탁드려요!

lewisseo91 committed:
feat(LottoNumbersDto): LottoNumbers view 분리, LottoNumber Comparable 추가

LottoNumbersGroupDto 추가, LottoNumbersDto 추가
- Dto 추가 후 view 에서 사용
LottoNumbers view 부분 method 분리
- toString addElementForMakingString convertSetToSortedLottoNumbersList 메서드
LottoNumber comparable 추가
- view 부분에서 사용하기 위한 메서드 추가
LottoMessage 삭제
- showMyLottoNumbers 메서드 message에서 삭제하고 LottoInputView 에서 데이터 가공하는 것으로 이동
LottoInputView 추가
- showMyLottoNumbers -> showMyLottoNumbersGroup 으로 이름 변경 후 showMyLottoNumbersGroup, printMyLottoNumbers 추가 기존 LottoNumbers에서 하던 로직 그대로 수행

lewisseo91 committed:
feat(LottoResultsDto): LottoResults view 분리

LottoResultsDto 추가
- Dto 추가 후 view 에서 사용
LottoResults view 부분 제거, method 분리
- toString 제거
LottoRankingStatus view 부분 제거
- view 부분 제거, 외부에서 prizeAmount 받기 위해 getter 추가
LottoNumbersTest 삭제
- 입력 관련 테스트 였기 때문에 제거
LottoMessage 삭제
- showResultsStats 삭제
LottoResultsView 추가
- LottoMessage view 부분 추가 showResultsStats, printLottoRankingStatus, getMatchDescription 추가

lewisseo91 committed:
fix(LottoNumbersGroup): view 부분 제거 (toString 메서드)

view 부분 제거 (toString 메서드)

lewisseo91 committed:
style(global): naming 제거(자료구조를 알 수 있는 List or String), 설명 부족한 naming 추가

자료구조를 알 수 있는 naming 제거(List or String), 설명 부족한 naming 추가
LottoResults
- lottoRankingAmountMap -. lottoRankingAmounts
LottoResult, LottoResultTest
- getResultRanking -> getResultRankingStatus
generateManualLottoNumberGroupTest
- getExampleLottoNumbersList -> getExampleLottoNumbers
- lottoNumbersList -> lottoNumbers
LottoNumbersGroup
- myLottoNumbersList -> myLottoNumbers
- lottoNumbersList -> resultLottoNumbers
- lottoResultsList -> lottoResults
LottoNumbers
- lottoNumbersString -> lottoNumbers
- splitLottoNumbersString -> splitLottoNumbers

lewisseo91 committed:
fix(LottoNumbers): shuffle 후 subList 이용 Set 작성

generateRandomLottoNumbers 내부 변경
- 리뷰어님 추천 받음! (꼭 나중에도 써먹자!)
getLottoNumbers
- getter 추가 (dto 가져오기 위함)

lewisseo91 committed:
fix(LottoResults): 수익률 덧셈 data type long 으로 변경

수익률 덧셈 data type long 으로 변경
- 1등이 20억이라 초과 가능성 있음

lewisseo91 committed:
fix(LottoController): 입력 계속 새로운 객체 입력 방식 변경

입력 계속 새로운 객체 입력 방식 변경
- scanner nextInt 후 nextLine 을 무조건 한번 더 받거나 Line으로 먼저 받고 Integer로 파싱 하는 방법 중에 후자 선택

lewisseo91 committed:
feat(LottoNumberDto): LottoNumberDto 추가

LottoNumberDto 추가, LottoInputView 에서 LottoNumberDto 로 map 과정 추가

lewisseo91 committed:
feat(LottoMessage): 예외 메시지 LottoMessage로 이동

예외 메시지 LottoMessage로 이동
- 예외 메시지도 뷰 관련 인 것 같아서 이동

lewisseo91 committed:
fix(LottoNumbersGroupTest): fixture로 만들어 줌

fixture로 만들어 줌

lewisseo91 commented:
안녕하세요!
꼼꼼한 리뷰 주셔서 정말 정말 감사합니다!
제 리뷰 요청 시에는 PR 메시지를 어디에 보내야 할 지 몰라서 여기에 적어서 코멘트를 남깁니다.

제가 DTO를 만들면서 기존에 너무 기계적으로 DTO를 생성 해 왔구나를 깨달았습니다.
DTO를 만들면서 제가 기존 객체에 필요한 요소들을 가져오기 위해서 getter를 가져왔는데
이 부분에 getter를 사용하는 것은 괜찮을까요?
혹시 getter를 쓰지 않고 dto를 생성하는 다른 방법이 있으면 알려주시면 정말 감사하겠습니다 ㅠㅠ

그리고 Fixture를 만들어 최대한 중복을 제거 해 보았는데 
static 안에서도 중복을 없애려고 최대한 시도해 보았습니다.
혹시나 static의 경우에 static으로 미리 등록해서 사용하는 경우, 다른 개발자가 본다면 더 헷갈릴 수도 있는 지 궁금합니다!

주말에도 피드백 주셔서 정말 감사합니다!!

nokchax commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -1,14 +1,13 @@
 import java.util.Objects;
 
-public class LottoNumber {
+public class LottoNumber implements Comparable<LottoNumber> {
```
👍

nokchax commented on src/main/java/lotto/dto/LottoNumbersDto.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.stream.Collectors;
+
+public class LottoNumbersDto {
+    private final Set<LottoNumber> lottoNumbers;
```
```suggestion
    private final Set<LottoNumberDto> lottoNumbers;
```
해당 도메인을 그대로 리턴하면 큰 의미가 없을거 같아요

nokchax replied:
혹은 그냥 LottoNumberDto를 없애고
```java
private final Set<Integer> lottoNumbers;
```
를 사용하는 방식도 있겠네요

lewisseo91 replied:
> 혹은 그냥 LottoNumberDto를 없애고
> 
> ```java
> private final Set<Integer> lottoNumbers;
> ```
> 
> 를 사용하는 방식도 있겠네요

아! 머리를 탁 쳤네요 감사합니다!

nokchax commented on src/main/java/lotto/dto/LottoNumbersGroupDto.java:
```diff
@@ -0,0 +1,18 @@
+import java.util.List;
+
+public class LottoNumbersGroupDto {
+    private final List<LottoNumbers> lottoNumbersGroup;
```
여기도
```suggestion
    private final List<LottoNumbersDto> lottoNumbersGroup;
```

nokchax commented on src/test/java/lotto/LottoNumbersTest.java:
```diff
@@ -15,42 +14,32 @@
 @DisplayName("로또 숫자들 테스트")
 public class LottoNumbersTest {
-    private LottoNumbers prizeLottoNumbers;
+    private static final LottoNumbers exampleLottoNumbers1 = new LottoNumbers(
```
질문에 대해 답을 드리자면
```
제가 DTO를 만들면서 기존에 너무 기계적으로 DTO를 생성 해 왔구나를 깨달았습니다.
DTO를 만들면서 제가 기존 객체에 필요한 요소들을 가져오기 위해서 getter를 가져왔는데
이 부분에 getter를 사용하는 것은 괜찮을까요?
```
A) 네 그정도는 괜찮다고 생각합니다. 그럼에도 getter를 없애라고 했던 이유는 데이터를 DTO로 넘기기 위해 getter를 만들어뒀는데 로직에 사용하게 되는 걸 방지하는 차원에서 말씀드린거예요.

```
혹시 getter를 쓰지 않고 dto를 생성하는 다른 방법이 있으면 알려주시면 정말 감사하겠습니다 ㅠㅠ
```
A) Domain 내부에서 그 Domain의 필드값을 DTO 생성자 내부로 직접 넘겨주고 객체를 생성하여 DTO 객체를 반환하는 방법이 있겠네요, 빌더 패턴을 사용해서 반환하던가요.
e.g.
``` java
public class LottoNumber implements Comparable<LottoNumber> {
    ...
    private final int number;
    ...
    public LottoNumberDto toDto() {
        return new LottoNumberDto(number);
    }
}
```
그런데 어떤 분들은 도메인 내부에 DTO에 종속적인 코드가 들어간다고 선호하지 않으시는 분들도 계셔요. 개인의 취향 차이라고 생각해요.
getter를 로직에서 활용만 안한다면 getter를 두어도 괜찮다고 생각합니다.(의도와는 다르게 어느새 getter를 사용하는게 문제지요ㅠ)

```
혹시나 static의 경우에 static으로 미리 등록해서 사용하는 경우, 다른 개발자가 본다면 더 헷갈릴 수도 있는 지 궁금합니다!
```
A) 저는 헷갈린다고 생각해 본적은 없는데, 만약 헷갈리신다거나 헷갈리는게 걱정 되신다면, 테스트 객체를 간편하게 만드는 헬퍼 클래스를 만드신 다던지, 그 객체에 생성자를 추가하시는 것도 방법입니다.

e.g.
```java
LottoNumbers example = new LottoNumbers(
        Stream.of(new LottoNumber(1), new LottoNumber(2), new LottoNumber(3),
                         new LottoNumber(4), new LottoNumber(5), new LottoNumber(6)
                   ).collect(Collectors.toSet()));
```
위와 같이 LottoNumbers 객체를 생성하기 위해서
```java
LottoNumbers example = Fixture.lottoNumbersOf(1, 2, 3, 4, 5, 6);
```
혹은
```java
LottoNumbers example = new LottoNumbers(1, 2, 3, 4, 5, 6);
```
다만 헬퍼 클래스를 사용하고 싶으시다면 test 패키지 내부에 아래와 같이 만들어줘야겠죠
```java
public class Fixture {
    private Fixture() {}
    
    public static LottoNumbers lottoNumbersOf(int... numbers) {
        return Arrays.stream(numbers)
                .mapToObj(LottoNumber::new)
                .collect(Collectors.collectingAndThen(Collectors.toSet(), LottoNumbers::new));
    }
}
```

lewisseo91 replied:
상세한 답변 정말정말 감사드립니다!

nokchax reviewed (changes requested):
동현님 열심이시군요, 빠르게 리뷰 반영해 주셨네요 💯 
남겨주신 질문이랑, DTO에 대해서 코멘트 달아드렸어요.
확인해보시고, 재요청 주세요 :)

lewisseo91 committed:
fix(LottoNumbersGroupDto): Dto 안에 domain 객체 dto 객체로 변경

LottoNumbersGroupDto, LottoNumbersDto, LottoNumberDto 변경, LottoInputView 에도 적용
- Dto 안에 domain 객체 dto 객체로 변경
- NumberDto 안에서 정렬이 되어야 해서 comparable 추가

lewisseo91 committed:
style(domain): domain 패키지 생성 후 이동

domain 패키지 생성 후 이동
- 나머지 수정 파일들은 import 로 인한 수정

lewisseo91 committed:
chore(domain): 필요없는 import 삭제

필요없는 import 삭제

lewisseo91 commented:
안녕하세요!
DTO 내부에 도메인 객체들을 DTO 객체로 변경하고 커밋하였습니다!

말씀해주신 빌더패턴은 정말 유용해서 꼭 잊지 말고 써야겠다고 생각했습니다!
그리고 Set<Integer>로 받는 것을 보았을 때, 'DTO 인데도 객체로 wrapping 되어야 한다는 강박이 있었구나' 라고 또 깨달았습니다.

이번에는 getter를 이용해 DTO 객체를 바꾸는 형태로 했지만 
이 부분 꼭 메모 해 두고 다음 step 이나 다음 미션에서 꼭 사용해 보겠습니다 ㅎㅎ
혹시나 이 부분 수정하기 원하신다면 바로 수정하겠습니다!

그리고 static 질문의 경우에는 제가 step3 진행하면서 해당 케이스들만 너무 오래보다 보니 
눈에 케이스들이 너무 익숙해져서 저에게만 편한 케이스인지 도저히 판단이 안 서서 도움을 요청하게 되었습니다. 
좋은 답변 감사합니다!
헷갈리지 않으시면 일단은 유지하고 다음 step 에서 Fixture 클래스를 활용 해 보겠습니다!

그리고 답변이 너무 좋아서 제 블로그에 기록하려고 하는데 혹시 괜찮을까요?

피드백 감사합니다!

nokchax commented:
블로그 기록은.. 네 ㅎㅎ.. 마음것 활용하셔도 좋습니다.

nokchax reviewed (approved):
동현님, 피드백 잘 반영해 주셨네요 👍 
다음 단계에서는 사용하지 않는 empty.txt는 삭제해주세요
수고하셨습니다, 머지할게요 다음 단계 진행해주세요 :)

lewisseo91 commented:
> 동현님, 피드백 잘 반영해 주셨네요 👍 다음 단계에서는 사용하지 않는 empty.txt는 삭제해주세요 수고하셨습니다, 머지할게요 다음 단계 진행해주세요 :)

감사합니다!
